python-test-samples/async-architectures/async-lambda-dynamodb/tests/integration/test_async_transform.py

```diff
@@ -0,0 +1,209 @@
+reside at the end of the process.
+'''
+def getStacks() -> str:
+
+    stack_name = os.environ.get("AWS_SAM_STACK_NAME")
+    if not stack_name:
+        raise Exception(
+            "Cannot find env var AWS_SAM_STACK_NAME. \n"
+            "Please include the stack name when running integration tests."
+        )
+
+    client = boto3.client("cloudformation")
+
+    try:
+        response = client.describe_stacks(StackName=stack_name)
+    except Exception as e:
+        raise Exception(
+            f"Cannot find stack {stack_name}. \n" f'Please make sure stack with the name "{stack_name}" exists.'
+        ) from e
+
+    return response["Stacks"]
+
+'''
+Put a lowercase string into the source bucket. 
+This the is beginning of the async process. 
+'''
+def put_object_into_source_bucket(unmodified_message, source_bucket_name, file_name):
+
+    print()
+
+    print("*** Putting object into S3... ")
+    client = boto3.client('s3')
+    response = client.put_object(Body=unmodified_message, Bucket=source_bucket_name, Key=file_name)
+    if response['ResponseMetadata']['HTTPStatusCode'] != HTTPStatus.OK:
+        raise f"Cannot put object into Source Bucket {source_bucket_name}"
+
+'''
+Poll to retrieve the modified string from a DynamoDB table at the end of the async process.
+'''    
+def test_retrieve_object_from_dynamodb(unmodified_message, modified_message, source_bucket_name, test_results_table, poll_timeout_duration_secs, test_filename):
+
+    print()
+    # set the maximum amount of time to poll before the test fails
+    # this can be based on our system's SLA
+    loop_start_time = datetime.now()
+    loop_max_time = loop_start_time + timedelta(seconds=poll_timeout_duration_secs)
+    global file_name
+
+    put_file_yet = False
+
+    while True:
+
+        if put_file_yet == False:
+            put_object_into_source_bucket(unmodified_message, source_bucket_name, file_name)
+            put_file_yet = True
+
+        # query dynamodb for the output
+        dynamodb = boto3.client("dynamodb")
+        response = dynamodb.get_item(
+            TableName=test_results_table,
+            Key={
+                'id': {'S': file_name}
+            }
+        )
+
+        # did we get the object?
+        try: 
+            x = response['Item']['message']
+            print("*** Properly transformed object found in DynamoDB. Success!")
+
+            # is the object formed correctly?
+            assert x['S'] == modified_message, f"Object malformed. Found {x}. Should be: {modified_message}"
+            
+            # we found the object and it's well formed, the test passed! let's exit the loop
+            break
+
+        except KeyError:
+            pass
+        
+        print("*** Querying DynamoDB for the object. Not found yet...")
+        
+        if (loop_max_time < datetime.now()):
+            assert False, ("*** Could not retrieve object from DynamoDB before timeout. Async process failed.")
```

Comment:
Will an `assert False` exit the test? Will that happen all the time or is there a way to configure pytest to run completely and collect all fails first? 

It may be a better practice to break after this, in the event you can configure pytest to not exit on failure. Otherwise this runs forever.

Reply:
I changed the implementation of the loop now using the backoff module.